src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorTagHelperBinderPhase.cs

```diff
@@ -188,5 +220,184 @@ private bool AssemblyContainsTagHelpers(string assemblyName, IReadOnlyList<TagHe
                 return false;
             }
         }
+
+        internal class ComponentDirectiveVisitor : DirectiveVisitor
+        {
+            private IReadOnlyList<TagHelperDescriptor> _tagHelpers;
+            private string _filePath;
+            private RazorSourceDocument _source;
+
+            public ComponentDirectiveVisitor(string filePath, IReadOnlyList<TagHelperDescriptor> tagHelpers, string currentNamespace)
+            {
+                _filePath = filePath;
+
+                // We don't want to consider non-component tag helpers in a component document.
+                _tagHelpers = tagHelpers.Where(t => ComponentMetadata.IsComponentTagHelperKind(t.Kind)).ToList();
+
+                for (var i = 0; i < _tagHelpers.Count; i++)
+                {
+                    var tagHelper = _tagHelpers[i];
+                    if (tagHelper.IsComponentFullyQualifiedNameMatch())
+                    {
+                        // If the component descriptor matches for a fully qualified name, using directives shouldn't matter.
+                        Matches.Add(tagHelper);
+                        continue;
+                    }
+
+                    var typeName = tagHelper.GetTypeName();
+                    if (tagHelper.IsChildContentTagHelper())
+                    {
+                        // If this is a child content tag helper, we want to add it if it's original type is in scope.
+                        // E.g, if the type name is `Test.MyComponent.ChildContent`, we want to add it if `Test.MyComponent` is in scope.
+                        var lastDot = typeName.LastIndexOf('.');
```

Comment:
Forgive my ignorance, ChildContent can never be generic an it?

Reply:
It can totally be generic. But that will not show in the tag helper type name. See https://github.com/aspnet/AspNetCore-Tooling/blob/master/src/Razor/src/Microsoft.CodeAnalysis.Razor/ComponentTagHelperDescriptorProvider.cs#L240